qemu/tests/balloon_hotplug.py

```diff
@@ -0,0 +1,86 @@
+
+    :param test:   QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env:    Dictionary with test environment.
+    """
+    def memory_balloon():
+        if params['os_type'] == 'windows':
+            balloon_test = balloon_check.BallooningTestWin(test, params, env)
+        else:
+            balloon_test = balloon_check.BallooningTestLinux(test, params, env)
+        for tag in params.objects('test_tags'):
+            params_tag = params.object_params(tag)
+            error_context.context("Running %s test" % tag, logging.info)
+            balloon_type = params_tag['balloon_type']
+            min_sz, max_sz = balloon_test.get_memory_boundary(balloon_type)
+            expect_mem = int(random.uniform(min_sz, max_sz))
+        quit_after_test = balloon_test.run_ballooning_test(expect_mem, tag)
+        if quit_after_test:
+            return
+        balloon_test.reset_memory()
+
+    pause = float(params.get("virtio_balloon_pause", 3.0))
+    idx = 0
+    err = ""
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    error_context.context("Hotplug and unplug balloon device in a loop",
+                          logging.info)
+    for i in xrange(int(params.get("balloon_repeats", 3))):
+        vm.devices.set_dirty()
+        new_dev = qdevices.QDevice("virtio-balloon-pci",
+                                   {'id': 'balloon%d' % idx},
+                                   parent_bus={'aobject': 'pci.0'})
+
+        error_context.context("Hotplug balloon device for %d times" % int(i+1),
+                              logging.info)
+        out = new_dev.hotplug(vm.monitor)
+        if out:
+            err += "\nHotplug monitor output: " + out
+        # Pause
+        time.sleep(pause)
+        ver_out = new_dev.verify_hotplug(out, vm.monitor)
+        if ver_out is False:
+            err += ("\nDevice is not in qtree %ss after hotplug:\n%s"
+                    % (pause, vm.monitor.info("qtree")))
+
+        error_context.context("Check whether balloon device work after hotplug",
+                              logging.info)
+        memory_balloon()
+
+        error_context.context("Unplug balloon device for %s times" % i+1,
+                              logging.info)
+        out = new_dev.unplug(vm.monitor)
+        if out:
+            err += "\nUnplug monitor output: " + out
+        # Pause
+        time.sleep(pause)
+        ver_out = new_dev.verify_unplug(out, vm.monitor)
+        if ver_out is False:
+            err += ("\nDevice is still in qtree %ss after unplug:\n%s"
+                    % (pause, vm.monitor.info("qtree")))
+
+        if err != "":
+            logging.error(vm.monitor.info("qtree"))
+            raise error_context.TestFail("Error occurred while hotpluging "
+                                         "virtio-pci. Iteration %s, monitor "
+                                         "output:%s" % (i, err))
```

Comment:
When the test failed, the log does not displayed completely:

  File "/root/test_home/qzhang/avocado-dir/usr/share/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/balloon_hotplug.py", line 84, in run
    raise error_context.TestFail("Error occurred while hotpluging "
AttributeError: 'module' object has no attribute 'TestFail'

Traceback (most recent call last):